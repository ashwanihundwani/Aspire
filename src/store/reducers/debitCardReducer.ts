
import { DEBIT_CARD_INFO_SUCCESS, UPDATE_DEBIT_CARD } from '../actions/types'

const initialState = {
  debitCard: null
};

export default (state = initialState, action: any) => {

  switch (action.type) {
    case DEBIT_CARD_INFO_SUCCESS: {

      return { ...state, debitCard: action.response };
    }
    case UPDATE_DEBIT_CARD: {

      return { ...state, debitCard: action.request };
    }
    default:
      return state;
  }
};
