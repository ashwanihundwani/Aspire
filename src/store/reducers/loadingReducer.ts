/**
 * Loading reducer made separate for easy blacklisting
 * Avoid data persist
 */

import { ENABLE_LOADER, DISABLE_LOADER } from '../actions/types';

const initialState = {
  isLoading: false,
};

export default (state = initialState, action: any) => {
  switch (action.type) {
    case ENABLE_LOADER: {
      return { ...state, isLoading: true };
    }
    case DISABLE_LOADER: {
      return { ...state, isLoading: false };
    }
    default:
      return state;
  }
}
