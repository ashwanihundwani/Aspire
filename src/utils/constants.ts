import { windowWidth, windowHeight } from "./utils";
import { Platform } from "react-native";

// Tab Name Constants
export const Tabs = { 
    home: "Home", 
    debit: "Debit Card", 
    payments: "Payments",
    credit: "Credit",
    account: "Profile"
} as const;

// Stack Route/Name Constants
export const StackScreens = { 
    debitCardTab: "Tab", 
    weeklySpentLimit: "WeeklySpendingLimit", 
    
} as const;

// Font Name Constants
export const FontNames = { 
    Avenir: "AvenirNext"  
} as const;

export const AppColors = {
    lightGray: "#DDDDDD",
    lightGreen: "#01D167",
    navigationBackColor: "#0C365A",
    navigationTintColor: "#FFFFFF",
    whiteColor: "#FFFFFF",
    darkTextColor: "#22222266",
    descriptionColor: "#222222",
    spentLimitText: "#22222233",
    separatorColor: "#DDDDDD",
    lightGreenWithOpacity: 'rgba(1,209,103, 0.07)',
    debitCardTitleColor: "#25345F",
    disabledSaveButton: "#EEEEEE"
}

export const CommonString = {
    currencySign: "S$",
    dollarSign: "$",
    switchSuffix:"_Switch"
}

export const DebitCardString = {
    headerTitle: "Debit Card",
    expiry: "Thru: ",
    cvv: "CVV: ",
    showCard: "Show card number",
    hideCard: "Hide card number",
    availableBalance: "Available Balance",
    spendLimitLabel: "Debit card spending limit",
    cardInfo: {
        cardNumberSplitter:"-",
        hiddenCVV: "* * *",
        dotString: '\u2022\u2022\u2022\u2022'
    }
}

export const WeeklySpendingString = {
    headerTitle: "Spending Limit",
    descriptionMessage: "Set a weekly debit card spending limit",
    infoMessage: "Here weekly means the last 7 days - not the calendar week", 
    saveButton: "Save",
    buttonTitle1: "S$ 5,000",
    buttonTitle2: "S$ 10,000",
    buttonTitle3: "S$ 20,000",
}

export const DebitCardOptions = {
    Titles : {
        topup: "Top-up account",
        weeklyspent: "Weekly spending limit",
        freezeCard: "Freeze card",
        newCard: "Get a new card",
        deactivated: "Deactivated cards"
    },
    Descriptions: {
        topup: "Deposit money to your account to use with card",
        weeklyspent: "You havenâ€™t set any spending limit on card",
        freezeCard: "Your debit card is currently active",
        newCard: "This deactivates your current debit card",
        deactivated: "Your previously deactivated cards"
    },
}

export const layouting = {
    // Standard margin calculation to accomodate various screen sizes accross devices.
    spacingFactor: 0.058 * windowWidth(),
}

export const testIDs = {
    showHideButton: "DebitCard.CardInfo.showHide",
    cardOption: "DebitCard.CardOption",
    spendProgress:"DebitCard.SpendProgress"
}

  