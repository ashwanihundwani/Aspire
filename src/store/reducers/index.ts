import { combineReducers } from 'redux';

import debitCardReducer from './debitCardReducer';
import loadingReducer from './loadingReducer';


export default combineReducers({
    debitCardReducer,
    loadingReducer
});