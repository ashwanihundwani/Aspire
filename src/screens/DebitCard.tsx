import React, { useState, useEffect, useRef } from 'react';
import { StyleSheet, Text, View, FlatList, Modal, Image, Animated, Platform } from 'react-native';
import { AppColors, DebitCardString, layouting, DebitCardOptions, CommonString, testIDs, ErrorMessages } from '../utils/constants';
import CardInfo from '../components/CardInfo';
import { avenirFont, FontStyle, windowHeight } from '../utils/utils';
import DebitCardOption, { CardOptionProps } from '../components/DebitCardOption';
import Images from '../../src/assets/images';
import CurrencyWithAmount from '../components/CurrencyWithAmount';
import { useDispatch, useSelector } from 'react-redux';
import { requestDebitCardInfo } from '../store/actions/debitCardActions';
import { DebitCardResponse } from '../models/DebitCardModel';
import Loader from '../components/Loader'

const spacingFactor = layouting.spacingFactor
const bottomSheetMaxOffset = 240;
const bottomSheetMinOffset = 0;
const bottomSheetScrollDistance = bottomSheetMaxOffset - bottomSheetMinOffset;

interface AvailableBalanceProps {
  balance: string
}

const AvailableBalance: React.FC<AvailableBalanceProps> = (props: AvailableBalanceProps) => {
  return (
    <View style={styles.balanceContainer}>
      <Text style={styles.balance}>{DebitCardString.availableBalance}</Text>
      <CurrencyWithAmount amount={props.balance} />
    </View>
  )
}
const DebitCard = (props: any) => {

  const dispatch = useDispatch();
  const debitCardResponse: any = useSelector((state: any): DebitCardResponse => state.debitCardReducer)
  const { debitCard, debitCardError } = debitCardResponse
  const loading: boolean = useSelector((state: any): boolean => state.loadingReducer.isLoading)
  const scrollY = useRef(new Animated.Value(0)).current;
  const [debitCardOptions, setDebitCardOptions] = useState<CardOptionProps[]>([
    {
      title: DebitCardOptions.Titles.topup,
      description: DebitCardOptions.Descriptions.topup,
      icon: Images.cardInfo.topup,
      showSwitch: false,
    },
    {
      title: DebitCardOptions.Titles.weeklyspent,
      description: DebitCardOptions.Descriptions.weeklyspent,
      icon: Images.cardInfo.weeklySpent,
      showSwitch: true,
      switchOn: false,
      switchEnabled: false
    },
    {
      title: DebitCardOptions.Titles.freezeCard,
      description: DebitCardOptions.Descriptions.freezeCard,
      icon: Images.cardInfo.freezeCard,
      showSwitch: true,
      switchOn: false
    },
    {
      title: DebitCardOptions.Titles.newCard,
      description: DebitCardOptions.Descriptions.topup,
      icon: Images.cardInfo.newCard,
      showSwitch: false,

    },
    {
      title: DebitCardOptions.Titles.deactivated,
      description: DebitCardOptions.Descriptions.topup,
      icon: Images.cardInfo.deactivateCard,
      showSwitch: false
    },
  ])

  const onSwitchValueChanged = (on: boolean, index: number) => {
    let tempOptions = [...debitCardOptions]
    tempOptions[index].switchOn = on
    setDebitCardOptions(tempOptions)
  }

  const onClick = (index: number) => {
    if (index === 1 && !debitCardOptions[index].switchOn) {
      props.navigation.navigate("WeeklySpendingLimit")
    }
  }

  const headerTranslateY = scrollY.interpolate({
    inputRange: [0, bottomSheetScrollDistance],
    outputRange: [0, -bottomSheetScrollDistance],
    extrapolate: 'clamp',
  });

  useEffect(() => {
    dispatch(requestDebitCardInfo());
    return () => {
      // Cleanup/unsuscribe from events
    }
  }, [])

  useEffect(()=> {
    if(debitCardError) {
      alert(ErrorMessages.networkError)
    }
  },[debitCardError])

  useEffect(()=> {
    if(debitCard && debitCard.weeklySpendLimit.length > 0) {
      let tempOptions = [...debitCardOptions]
      tempOptions[1].switchOn = true
      tempOptions[1].switchEnabled = true
      tempOptions[1].description = DebitCardOptions.Descriptions.weeklySpendSaved + debitCard.weeklySpendLimit
      setDebitCardOptions(tempOptions)
    }
  },[debitCard])

  

  return (
    <View style={{ flex: 1, backgroundColor: AppColors.navigationBackColor }}>
      {debitCardResponse && debitCard &&
        <>
          <View style={styles.logo}>
            <Text style={styles.headerTitle}>{DebitCardString.headerTitle}</Text>
            <Image style={{ marginRight: layouting.spacingFactor, marginTop: 0 }} source={Images.app.logo}></Image>
          </View>
          <AvailableBalance balance={debitCard.balance} />
          <Animated.View style={[styles.bottomSheet, { transform: [{ translateY: headerTranslateY }] }]} />
          <Animated.FlatList testID="dddddd"
            onScroll={Animated.event(
              [{ nativeEvent: { contentOffset: { y: scrollY } } }], // event.nativeEvent.contentOffset.x to scrollX
              { useNativeDriver: true }, // use native driver for animation
            )}
            style={styles.optionsList}
            bounces={false}
            showsVerticalScrollIndicator={false}
            ListHeaderComponent={
              <View style={styles.cardInfo}>
                <CardInfo
                  showProgress={debitCardOptions[1].switchOn ? debitCardOptions[1].switchOn : false}
                  holderName={debitCard.holderName}
                  cardNumber={debitCard.cardNumber}
                  expiry={debitCard.expiry}
                  CVV={debitCard.cvv}
                  weeklySpentLimit={debitCard.weeklySpendLimit}
                  amountSpent={debitCard.spentAmount}
                />
              </View>
            }
            data={debitCardOptions}
            renderItem={({ item, index }) => {
              const { title, description, icon, showSwitch, switchOn, switchEnabled } = item
              return <DebitCardOption
                title={title}
                description={description}
                icon={icon}
                showSwitch={showSwitch}
                switchEnabled={switchEnabled}
                switchOn={switchOn}
                index={index}
                onSwitchValueChanged={onSwitchValueChanged}
                onClick={onClick}
                id={testIDs.cardOption + index}
              />
            }}>
          </Animated.FlatList>
        </>
      }
      <Loader isLoading={loading} />
    </View>
  );
};

const styles = StyleSheet.create({
  logo: {
    flexDirection: "row",
    marginTop: 50,
    justifyContent: "space-between"
  },
  headerTitle: {
    marginTop: Platform.OS === 'ios'? 20 : 16,
    fontFamily: avenirFont(FontStyle.Bold),
    fontSize: 24,
    color: AppColors.navigationTintColor,
    marginLeft: 24
  },
  balanceContainer: {
    marginTop: spacingFactor
  },
  balance: {
    fontSize: 14,
    color: AppColors.whiteColor,
    marginLeft: spacingFactor,
    fontFamily: avenirFont(FontStyle.Medium),
  },
  balanceContent: {
    flexDirection: "row",
    marginLeft: spacingFactor,
    marginTop: 10,
  },
  optionsList: {
    position: "absolute",
    top: 0,
    bottom: 2
  },
  cardInfo: {
    paddingTop: 150,
    backgroundColor: "transparent"
  },
  bottomSheet: {
    position: "absolute",
    top: 283,
    height: windowHeight() - 50,
    alignSelf: "flex-end",
    borderTopLeftRadius: spacingFactor,
    borderTopRightRadius: spacingFactor,
    backgroundColor: "white",
    width: "100%"
  }
});

export default DebitCard

