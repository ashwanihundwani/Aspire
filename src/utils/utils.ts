
import { FontNames } from './constants'
import { Dimensions } from 'react-native'

export enum FontStyle {
    Regular = "Regular",
    Medium = "Medium",
    DemiBold = "DemiBold",
    Bold = "Bold",
}

export function avenirFont(style: FontStyle) {
    return FontNames.Avenir + "-" + style
}

export function windowWidth() {
    return Dimensions.get("window").width
}

export function windowHeight() {
    return Dimensions.get("window").height
}

export function parseNumberFromString(str: string) {
    return parseInt(str.replace(/,/g, ''), 10);
}