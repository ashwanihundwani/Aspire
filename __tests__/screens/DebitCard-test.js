/**
 * @format
 */

import 'react-native';
import React from 'react';
import DebitCard from '../../src/screens/DebitCard';
import renderer from 'react-test-renderer';

import { Provider, useStore } from 'react-redux';
import reducers from '../../src/store/reducers'
import {cleanup,render,  fireEvent,} from '@testing-library/react-native'
import { testIDs, CommonString } from '../../src/utils/constants';

// Debit card
// 1. Click show hide button => expect Card Number & CVV to be hidden (Done)
// 2. Click Weekly spend toggle => expect Progressbar to be hidden. (In Progress)
// 3. Click Weekly spend row => expect WeeklySpend screen to appear.

//WeeklySpentAmount:
// 1. Click all 5000, 10000, 20000 buttons and check if amount label is updated.
// 2. Check Save button is disable if amount label is blank
// 3. Check Save button is enabled if amount label is filled
// 4. Click Save button => Debit Card screen should show up.
// 5. Check Weekly spend limit is updated on Debit Card screen.


it('renders correctly', () => {
    const tree = renderer
      .create(<Provider store={store}><DebitCard /></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
});

// it("show/hide Spend Progress on switch button toggle", () => {
//   const store = useStore().
//   const { getByTestId } = render(<Provider store={store}><DebitCard /></Provider>)

//   const testId  = "touchable"
//   getByTestId("dddddd")
//   //fireEvent.press(getByTestId(testId))

//   //expect(getByTestId(testIDs.spendProgress)).toBeFalsy()
// })

afterEach(cleanup)
