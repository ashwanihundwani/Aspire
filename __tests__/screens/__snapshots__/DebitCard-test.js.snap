// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#0C365A",
      "flex": 1,
    }
  }
/>
`;
